subsys/fs/ext2/ext2_ops.c

```diff
@@ -0,0 +1,712 @@
+/*
+ * Copyright (c) 2023 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <zephyr/init.h>
+#include <zephyr/fs/fs.h>
+#include <zephyr/fs/fs_sys.h>
+#include <zephyr/fs/ext2.h>
+#include <zephyr/logging/log.h>
+
+#include "../fs_impl.h"
+#include "ext2.h"
+#include "ext2_impl.h"
+#include "ext2_struct.h"
+
+LOG_MODULE_DECLARE(ext2);
+
+/* File operations */
+
+static int ext2_open(struct fs_file_t *filp, const char *fs_path, fs_mode_t flags)
+{
+	int rc, ret = 0;
+	struct ext2_data *fs = filp->mp->fs_data;
+
+	if (!fs) {
+		return -EINVAL;
+	}
+
+	LOG_DBG("Open mode: Rd:%d Wr:%d App:%d Creat:%d",
+			(flags & FS_O_READ) != 0,
+			(flags & FS_O_WRITE) != 0,
+			(flags & FS_O_APPEND) != 0,
+			(flags & FS_O_CREATE) != 0);
+
+	const char *path = fs_impl_strip_prefix(fs_path, filp->mp);
+	struct ext2_lookup_args args = {
+		.path = path,
+		.inode = NULL,
+		.flags = LOOKUP_ARG_OPEN,
+	};
+
+	if (flags & FS_O_CREATE) {
+		args.flags |= LOOKUP_ARG_CREATE;
+		args.parent = NULL;
+	}
+
+	rc = ext2_lookup_inode(fs, &args);
+	if (rc < 0) {
+		return rc;
+	}
+
+	/* Inodes allocated by lookup. Must be freed in manually. */
+	struct ext2_inode *found_inode = args.inode;
+
+	/* Not NULL iff FS_O_CREATE and found_inode == NULL */
+	struct ext2_inode *parent = args.parent;
+
+	/* File has to be created */
+	if (flags & FS_O_CREATE && found_inode == NULL) {
+		LOG_DBG("Returned from lookup & create: '%s':%d creating file: %d",
+				path + args.name_pos, args.name_len, found_inode == NULL);
+
+		struct ext2_inode *new_inode;
+
+		rc = ext2_inode_get(fs, 0, &new_inode);
+		if (rc < 0) {
+			ret = rc;
+			goto out;
+		}
+
+		rc = ext2_create_file(parent, new_inode, &args);
+		if (rc < 0) {
+			ext2_inode_drop(new_inode);
+			ret = rc;
+			goto out;
+		}
+
+		found_inode = new_inode;
+	}
+
+	if ((found_inode->i_mode & EXT2_S_IFMT) != EXT2_S_IFREG) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	struct ext2_file *f = ext2_heap_alloc(sizeof(struct ext2_file));
+
+	if (!f) {
+		ret = -ENOMEM;
+		goto out;
+	}
+
+	f->f_inode = found_inode;
+	f->f_off = 0;
+	f->f_flags = flags & (FS_O_RDWR | FS_O_APPEND);
+
+	filp->filep = f;
+
+	ext2_inode_drop(parent);
+	return 0;
+
+out:
+	ext2_inode_drop(found_inode);
+	ext2_inode_drop(parent);
+	return ret;
+}
+
+static int ext2_close(struct fs_file_t *filp)
+{
+	int rc;
+	struct ext2_file *f = filp->filep;
+
+	if (!f) {
+		return 0;
+	}
+
+	rc = ext2_inode_sync(f->f_inode);
+	if (rc < 0) {
+		goto out;
+	}
+
+	rc = ext2_inode_drop(f->f_inode);
+	if (rc < 0) {
+		goto out;
+	}
+
+	ext2_heap_free(f);
+out:
+	return rc;
+}
+
+static ssize_t ext2_read(struct fs_file_t *filp, void *dest, size_t nbytes)
+{
+	struct ext2_file *f = filp->filep;
+
+	if (!f) {
+		return -EBADF;
+	}
```

Comment:
It is possible when user provides an invalid file structure.

Reply:
OK, so user can provide invalid `filp`, we do not check for; also NULL is only one possible invalid in the 32bit range of possible.

Reply:
It is hard to predict what user provides as param to function call. The `flip` is supposed to be valid object, but we do not have control on what user provides here as file system objects are provided from user space. So we can get any crap here, and it has very bit changes of not being NULL, which means that it would be hard to figure out whether it has been initialized or correct.
Btw. if you have some NULL checks that should not trigger, generally, you can use CHECK_IF macros with them, which are only enabled with Kconfig options. Unfortunately there is no granularity so either everything in CHECK_IF is tested or yolo.

Reply:
removed